provider/apple.go

```diff
@@ -0,0 +1,532 @@
+type LoadFromFileFunc struct {
+	Path string
+}
+
+// LoadApplePrivateKeyFromFile return instance for pre-defined loader function from local file
+func LoadApplePrivateKeyFromFile(path string) LoadFromFileFunc {
+	return LoadFromFileFunc{
+		Path: path,
+	}
+}
+
+// LoadPrivateKey implement pre-defined (built-in) PrivateKeyLoaderInterface interface method for load private key from local file
+func (lf LoadFromFileFunc) LoadPrivateKey() ([]byte, error) {
+	if lf.Path == "" {
+		return nil, errors.New("empty private key path not allowed")
+	}
+
+	kFile, err := os.Open(lf.Path)
+	if err != nil {
+		return nil, err
+	}
+	keyValue, err := ioutil.ReadAll(kFile)
+	if err != nil {
+		return nil, err
+	}
+	err = kFile.Close()
+	return keyValue, err
+}
+
+// NewApple create new AppleProvider instance with custom name and parameters
+// endpoints used in constructor  required only for testing and can't be define one outside the package.
+// Private key must be set, when instance create call, for create `client_secret`
+func NewApple(name string, p Params, appleCfg AppleConfig, endpoints oauth2.Endpoint, privateKeyLoader PrivateKeyLoaderInterface) (*AppleHandler, error) {
+
+	if name == "" {
+		return nil, errors.New("empty name for apple provider not allowed")
+	}
+
+	if p.L == nil {
+		p.L = logger.NoOp
+	}
+
+	// configuring default values
+	if appleCfg.UserAgent == "" {
+		appleCfg.UserAgent = defaultUserAgent
+	}
+	if endpoints.AuthURL == "" {
+		endpoints.AuthURL = appleAuthURL
+	}
+	if endpoints.TokenURL == "" {
+		endpoints.TokenURL = appleTokenURL
+	}
+
+	ah := AppleHandler{
+		Params: p,
+		name:   name,
+
+		conf: AppleConfig{
+			ClientID:           appleCfg.ClientID,
+			TeamID:             appleCfg.TeamID,
+			KeyID:              appleCfg.KeyID,
+			ClientSecretExpire: appleCfg.ClientSecretExpire,
+			UserAgent:          appleCfg.UserAgent,
+			Scopes:             appleCfg.Scopes,
+		},
+
+		endpoint: endpoints,
+		mapUser: func(claims jwt.MapClaims) token.User {
+			var usr token.User
+			if uid, ok := claims["sub"]; ok {
+				usr.ID = fmt.Sprintf("apple_%s", uid.(string))
+			}
+
+			// this claim exist if "email" defined in scope
+			if email, ok := claims["email"]; ok {
+				usr.Email = email.(string)
+			}
+			if emailVerified, ok := claims["email_verified"]; ok {
+				usr.SetBoolAttr("email_verified", emailVerified.(string) == "true")
+			}
+
+			return usr
+		},
+	}
+
+	if privateKeyLoader == nil {
+		return nil, errors.New("private key loader function undefined")
+	}
+	ah.PrivateKeyLoader = privateKeyLoader
+
+	err := ah.initPrivateKey()
+	return &ah, err
+}
+
+// initPrivateKey parse Apple private key and assign to AppleHandler
+func (ah *AppleHandler) initPrivateKey() error {
+	if ah.PrivateKeyLoader == nil {
+		return errors.New("private key loader interface is nil")
+	}
+	if ah.conf.KeyID == "" {
+		return errors.New("keyID can't be empty")
+	}
+	sKey, err := ah.PrivateKeyLoader.LoadPrivateKey()
+	if err != nil {
+		return err
```

Comment:
Better wrap it with some context, like "problem loading private key", as well as others in that function.

Reply:
added error wrap in current in commit [8c4df02 ](https://github.com/go-pkgz/auth/pull/86/commits/8c4df026d051c2a4d6a2260156917e2eb418b92a)